### PR TITLE
[Issue:103] Document the use of a "personality" module with the api traitlet

### PR DESCRIPTION
**Pull request by nitind**
Adds a section for "Other modes" to the generated docs, explaining.

(c) Copyright IBM Corp. 2016

### CODE REVIEW

**Comment by nitind** [comment]
For https://github.com/jupyter/kernel_gateway/issues/103

File: docs/source/http-mode.md [hunk 1]
```diff
@@ -1,8 +1,6 @@
 ## `notebook-http` Mode
 
-**NOTE** This mode may one day become a plug-in to the Jupyter Kernel Gateway or split into a separate project. The concepts documented here, however, will survive any move.
-
-The `KernelGatewayApp.api` command line argument can be set to `notebook-http`. In this mode, the kernel gateway exposes annotated cells in the `KernelGatewayApp.seed_uri` notebook as HTTP resources.
+The `KernelGatewayApp.api` command line argument can be set to `kernel_gateway.notebook_http`. This personality has the kernel gateway expose annotated cells in the `KernelGatewayApp.seed_uri` notebook as HTTP resources.
```
Comment (parente) [inline docs/source/http-mode.md]: Maybe the indicate that "personality" and "mode" are synonymous:

"This mode, or *personality*, has the kernel gateway expose annotated cells in the `KernelGatewayApp.seed_uri` notebook as HTTP resources."

File: docs/source/plug-in.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
+## Other modes
```
Comment (parente) [inline docs/source/plug-in.md]: Maybe "Developing Modes"?

File: docs/source/plug-in.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+The `KernelGatewayApp.api` can be set to the name of any public module supplying a personality. This allows for suporting alternate kernel communications mechanisms.
```
Comment (parente) [inline docs/source/plug-in.md]: "... of any module in the Python path"

File: docs/source/plug-in.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+### Factory function
+`create_personality(*args, **kwargs)`
```
Comment (parente) [inline docs/source/plug-in.md]: You should name the `parent` argument here, even though the implementation does **kwargs as a passthrough.

File: docs/source/plug-in.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+The only function required in the module, it is passed a named argument `parent` whose value is the kernel gateway application. The function must return an object implementing the following lifecycle and delegated functions:
+
+### Lifecycle functions
```
Comment (parente) [inline docs/source/plug-in.md]: A couple things:

- The doc should recommend that the class derive from LoggingConfigurable.
- It's not clear that the lifecycle nor delegated functions are actually methods on the instance returned. 

I think splitting up the sections will make it hard for readers to envision what exactly they need to implement. Maybe instead you could show a complete, empty module template with docstrings on the functions describing what they need to do, like:

```
from traitlets.config.configurable import LoggingConfigurable

class TemplatePersonality(LoggingConfigurable):
    def init_configurables(self):
        """Put docstring here."""
        pass 

    def shutdown(self):
        """Put docstring here."""
        pass 

    def create_request_handlers(self):
        """Put docstring here."""
        pass 

    def should_seed_cell(self, code):
        """Put docstring here."""
        pass

def create_personality(self, parent):
    return TemplatePersonality(parent=parent)
```

Comment (parente) [inline docs/source/plug-in.md]: After the template, relative links to the two files implementing the jupyter-websocket and notebook-http modes / personalities would be helpful.

File: kernel_gateway/jupyter_websocket/__init__.py [hunk 6]
```diff
@@ -50,7 +50,7 @@ def create_request_handlers(self):
             handlers.append(new_handler)
         return handlers
 
-    def should_seed_cell(self, code):
+    def should_seed_cell(self, cell):
```
Comment (parente) [inline kernel_gateway/jupyter_websocket/__init__.py]: Is it a code string or a cell object? The docstring should have a `Parameters` section describing the param and its type plus the return value of the method. (Same goes for the template above if you adopt it and the other code -> cell change below.)

File: docs/source/config-options.md [hunk 7]
```diff
@@ -36,18 +36,19 @@ KernelGatewayApp options
 --KernelGatewayApp.allow_notebook_download=<Bool>
     Default: False
     Optional API to download the notebook source code in notebook-http mode,
-    defaults to not allow
+    defaults to not allow (KG_ALLOW_NOTEBOOK_DOWNLOAD env var)
```
Comment (parente) [inline docs/source/config-options.md]: 👍  for fixing

**Comment by parente** [comment]
Good updates! 🍰

**Pull request merged**